msgpack/unpack.h

```diff
@@ -342,32 +342,53 @@ static int unpack_callback_ext(unpack_user* u, const char* base, const char* pos
             py = PyObject_CallFunction(u->timestamp_t, "(Lk)", ts.tv_sec, ts.tv_nsec);
         }
         else { // float or datetime
-            PyObject *a = PyFloat_FromDouble((double)ts.tv_nsec);
-            if (a == NULL) return -1;
+            if (u->timestamp == 3) {  // datetime
+                // Calculate datetime using epoch + delta 
+                // due to limitations PyDateTime_FromTimestamp on Windows with negative timestamps
+                PyObject *t0 = PyLong_FromLongLong(0L);
+                if (t0 == NULL) {
+                    return -1;
+                }
 
-            PyObject *b = PyNumber_TrueDivide(a, u->giga);
-            Py_DECREF(a);
-            if (b == NULL) return -1;
+                PyObject *t = PyTuple_Pack(2, t0, u->utc);
+                Py_DECREF(t0);
+                if (t == NULL) {
+                    return -1;
+                }
 
-            PyObject *c = PyLong_FromLongLong(ts.tv_sec);
-            if (c == NULL) {
-                Py_DECREF(b);
-                return -1;
-            }
+                PyObject *t_epoch = PyDateTime_FromTimestamp(t);
```

Comment:
```c
    PyObject *epoch = PyDateTimeAPI->DateTime_FromDateAndTimeAndFold(1970, 1, 1, 0, 0, 0, 0, u->utc, 0, PyDateTimeAPI->DateTimeType);
```

Reply:
Done

Reply:
It seems the proposed method is not available in Python 3.5

We don't need the "...AndFold" anyways, I could use "PyDateTime_FromDateAndTime"